Prevent attempts to load null values using the cursor.getFoo(index) meth...

mcumings opened the pull request:
Prevent attempts to load null values using the cursor.getFoo(index) method calls on nullable, non-primitive fields.

The new behavior is to detect nullable field values and wrap their loading behavior in a Cursor.isNull(index) check.  This allows for a distinction between values which are not present and those which have their default (Java) value.

Also fixed a potential test escape and added more granular resolution to BoxedValuesTest.

mcumings commented:
Not sure how you're managing develop versus master.  Is this change better suited to a merge to master?

agrosner commented on library/src/androidTest/java/com/raizlabs/android/dbflow/test/sql/BoxedModel.java:
```diff
@@ -6,26 +6,63 @@
+    @Column(notNull = true)
+    public Short shortFieldNotNull = 1;
+
+    @Column()
```
remove paranthesis

agrosner commented on library/src/androidTest/java/com/raizlabs/android/dbflow/test/sql/BoxedModel.java:
```diff
@@ -6,26 +6,63 @@
+    @Column(notNull = true)
+    public Integer integerFieldNotNull = 1;
+
+    @Column()
```
same as above

agrosner commented on library/src/androidTest/java/com/raizlabs/android/dbflow/test/sql/BoxedModel.java:
```diff
@@ -6,26 +6,63 @@
+    @Column(notNull = true)
+    public Long longFieldNotNull = 1L;
+
+    @Column()
```
same as above

agrosner commented on library/src/androidTest/java/com/raizlabs/android/dbflow/test/sql/BoxedModel.java:
```diff
@@ -6,26 +6,63 @@
+    @Column(notNull = true)
+    public Float floatFieldNotNull = 1.0f;
+
+    @Column()
```
same as above

agrosner commented on library/src/androidTest/java/com/raizlabs/android/dbflow/test/sql/BoxedModel.java:
```diff
@@ -6,26 +6,63 @@
-    public String station;
+    public Double doubleFieldNotNull = 1.0;
+
+    @Column()
```
same as above

agrosner commented on library/src/androidTest/java/com/raizlabs/android/dbflow/test/sql/BoxedModel.java:
```diff
@@ -6,26 +6,63 @@
-    public String stand;
+    public String stringFieldNotNull = "1";
+
+    @Column()
```
same as above.

agrosner commented on library/src/androidTest/java/com/raizlabs/android/dbflow/test/sql/BoxedValueTest.java:
```diff
@@ -3,24 +3,129 @@
+        testObject.id = SEQUENCE_ID.getAndIncrement();
+    }
+
+    public void testBoxedValues_nullId() {
```
dont believe we need to test every field here to see if its saveable or not. rather all we need to test is a non-null and normal field can be saved.

agrosner commented on compiler/src/main/java/com/raizlabs/android/dbflow/processor/utils/ModelUtils.java:
```diff
@@ -137,21 +138,36 @@ public static void writeLoadFromCursorDefinitionField(JavaWriter javaWriter, Pro
+        javaWriter.beginControlFlow("if (%1s != -1) ", index);
 
         String cursorStatment = ModelUtils.getCursorStatement(newFieldType, columnName);
+        AdapterQueryBuilder nullableQueryBuilder = new AdapterQueryBuilder(queryBuilder.getQuery());
```
can you write this without using a second ```AdapterQueryBuilder```?

agrosner commented:
@mcumings looks pretty good. Just a few comments, but also this PR is not mergeable. Please fix! But yes all PRs should go against develop.

mcumings commented:
I had added all the individual cases to BoxedValueTest to dd coverage around com.raizlabs.android.dbflow.processor.writer.LoadCursorWriter#CURSOR_METHOD_MAP but have removed them as requested.

Also updated the ModelUtils logic to extract the assignment logic into a separate method containing a single builder.  I like this approach better anyhow so thanks for that. :)

agrosner commented:
@mcumings thanks for your help! I am always happy to accept pull requests.

mcumings commented:
No sweat, good sir.

agrosner commented:
@mcumings adding you as a contributor at the bottom of the README!
